### PR TITLE
DatasetSplitter

### PR DESCRIPTION
**Pull request by ngreenwald**
This PR creates a DatasetSplitter class. This creates sequentially increasing subsets of a data_dict to enable comparisons of accuracy with increasing dataset size. Each larger split contains all of the data in the smaller splits, plus whatever incremental additional samples are needed for the specified size. This ensures that all of the data in previous splits is seen in each subsequent split.

### CODE REVIEW

File: caliban_toolbox/dataset_splitter.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                                         'split {} is {}'.format(idx, splits[idx],
+                                                                 idx + 1, splits[idx + 1]))
+            if splits[-1] > 1:
+                raise ValueError('Largest split cannot be greater than 1, '
+                                 'got {}'.format(splits[-1]))
+            self.splits = splits
+
+    def _validate_dict(self, train_dict):
+        if 'X' not in train_dict.keys() or 'y' not in train_dict.keys():
+            raise ValueError('X and y must be keys in the training dictionary')
```
Comment (willgraf) [inline caliban_toolbox/dataset_splitter.py:61]: Just check if the key is in the dictionary, not if its in the keys.  `keys()` creates a new list and the lookup goes from O(1) to O(n).

```suggestion
        if 'X' not in train_dict or 'y' not in train_dict:
            raise ValueError('X and y must be keys in the training dictionary')
```

File: caliban_toolbox/dataset_splitter.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            self.splits = [0.05, 0.10, 0.25, 0.5, 0.75, 1]
+        else:
+            if splits[0] <= 0:
+                raise ValueError('Smallest split must be non-zero, got {}'.format(splits[0]))
+            else:
+                for idx in range(len(splits) - 1):
+                    if splits[idx] >= splits[idx + 1]:
+                        raise ValueError('Splits must be increasing in size. Split {} is {}, '
+                                         'split {} is {}'.format(idx, splits[idx],
+                                                                 idx + 1, splits[idx + 1]))
```
Comment (willgraf) [inline caliban_toolbox/dataset_splitter.py:53]: Why not just sort the list of splits?

File: caliban_toolbox/dataset_splitter.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        index = np.arange(N_batches)
+        permuted_index = np.random.RandomState(seed=self.seed).permutation(index)
+        split_dict = {}
+        for split in self.splits:
+            new_train_dict = {}
+            train_size = int(split * N_batches)
+            split_idx = permuted_index[0:train_size]
+            new_train_dict['X'] = X[split_idx]
+            new_train_dict['y'] = y[split_idx]
+            split_dict[str(split)] = new_train_dict
```
Comment (willgraf) [inline caliban_toolbox/dataset_splitter.py:77]: Why cast this to a string?

**Review by willgraf — CHANGES_REQUESTED** [review]
Why does this need to create so many splits?

Why not just have a function that splits the data according to a single split value?

**Comment by ngreenwald** [comment]
The reason we have all of these splits is because we want to use this to assess how accuracy changes with dataset size. So we specify in advance all of the different splits of the dataset we want, and then train a model on each for benchmarking purposes.

File: caliban_toolbox/dataset_splitter_test.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    assert ds.seed == seed
+    assert ds.splits == splits
+
+    # unsorted splits get sorted
+    splits = [0.8, 0.3, 0.5]
+    ds = DatasetSplitter(seed=seed, splits=splits)
+    splits.sort()
+
+    assert ds.splits == splits
```
Comment (willgraf) [inline caliban_toolbox/dataset_splitter_test.py:47]: `splits` is the same list as `ds.splits`, and this test will pass if you comment out the `splits.sort()`. Instead, maybe you can make an assertion about the elements of the list.

**Review by willgraf — CHANGES_REQUESTED** [review]

**Comment by ngreenwald** [comment]
The reason Dave initially casted the keys of the dict to strings was so that the dict can be saved as NPZ file. NPZ doesn't allow float keys, so I switched it back to strings in this version.

File: caliban_toolbox/dataset_splitter.py (hunk 3 above)
Comment (willgraf) [reply caliban_toolbox/dataset_splitter.py:75]: As mentioned elsewhere, it is cast to a string for JSON format.

**Review by willgraf — APPROVED** [review]

**Pull request merged**